File: backend/app/schemas/rectification.py
from datetime import date, datetime

from pydantic import BaseModel, Field


class RectificationBase(BaseModel):
    batch_id: int | None = None
    discipline: str | None = None
    building: str | None = None
    floor: str | None = None
    system_name: str | None = None
    task_name: str | None = None
    issue_description: str | None = None
    delay_level: str | None = None
    responsible_person: str | None = None
    responsible_unit: str | None = None
    planned_finish_date: date | None = None
    status: str = "open"
    review_result: str | None = None
    remark: str | None = None


class RectificationCreate(RectificationBase):
    source_type: str = "manual"


class RectificationUpdate(BaseModel):
    responsible_person: str | None = None
    responsible_unit: str | None = None
    planned_finish_date: date | None = None
    status: str | None = None
    review_result: str | None = None
    remark: str | None = None


class RectificationStatusUpdate(BaseModel):
    status: str
    remark: str | None = None


class RectificationFromDelayedItem(BaseModel):
    batch_id: int
    progress_item_id: int


class RectificationFromWarningRecord(BaseModel):
    warning_record_id: int


class RectificationRead(BaseModel):
    id: int
    project_id: int
    batch_id: int | None = None
    source_batch_label: str | None = None
    source_baseline_plan_id: int | None = None
    source_baseline_plan_name: str | None = None
    progress_item_id: int | None = None
    warning_record_id: int | None = None
    source_type: str
    source_id: int | None = None
    source_label: str
    discipline: str | None = None
    building: str | None = None
    floor: str | None = None
    system_name: str | None = None
    task_name: str | None = None
    issue_description: str | None = None
    delay_level: str | None = None
    delay_level_label: str
    actual_percent: float | None = None
    planned_percent: float | None = None
    progress_deviation: float | None = None
    responsible_person: str | None = None
    responsible_unit: str | None = None
    planned_finish_date: date | None = None
    status: str
    status_label: str
    review_result: str | None = None
    remark: str | None = None
    is_overdue: bool
    created_at: datetime
    updated_at: datetime
    closed_at: datetime | None = None

    model_config = {"from_attributes": True}


class RectificationListResponse(BaseModel):
    items: list[RectificationRead]
    total: int
    page: int = 1
    page_size: int = 20


class RectificationSummary(BaseModel):
    total: int = 0
    open: int = 0
    in_progress: int = 0
    completed: int = 0
    closed: int = 0
    ignored: int = 0
    overdue: int = 0
    serious: int = 0
    new_this_week: int = 0
    closed_this_week: int = 0


class RectificationCreateResponse(BaseModel):
    item: RectificationRead
    created: bool
    message: str


class RectificationExportFilters(BaseModel):
    status: str | None = Field(default=None)
    discipline: str | None = None
    building: str | None = None
    floor: str | None = None
    delay_level: str | None = None
    keyword: str | None = None


class RectificationBatchUpdate(BaseModel):
    ids: list[int]
    status: str | None = None
    responsible_person: str | None = None
    responsible_unit: str | None = None
    planned_finish_date: date | None = None
    remark: str | None = None


class RectificationBatchUpdateResponse(BaseModel):
    updated_count: int
    skipped_count: int
    skipped_ids: list[int] = []


class RectificationFilterOptions(BaseModel):
    disciplines: list[str] = []
    buildings: list[str] = []
    floors: list[str] = []
    responsible_persons: list[str] = []
    responsible_units: list[str] = []
    delay_levels: list[str] = []
    statuses: list[str] = []
    source_types: list[str] = []
    floors_by_building: dict[str, list[str]] = {}


class RectificationActionLogRead(BaseModel):
    id: int
    rectification_item_id: int
    project_id: int
    action: str
    action_label: str
    operator: str = "本地用户"
    from_status: str | None = None
    from_status_label: str | None = None
    to_status: str | None = None
    to_status_label: str | None = None
    content: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/rectification.ts
export type RectificationStatus = 'open' | 'in_progress' | 'completed' | 'closed' | 'ignored'

export type RectificationItem = {
  id: number
  project_id: number
  batch_id?: number | null
  source_batch_label?: string | null
  source_baseline_plan_id?: number | null
  source_baseline_plan_name?: string | null
  progress_item_id?: number | null
  warning_record_id?: number | null
  source_type: string
  source_id?: number | null
  source_label: string
  discipline?: string | null
  building?: string | null
  floor?: string | null
  system_name?: string | null
  task_name?: string | null
  issue_description?: string | null
  delay_level?: string | null
  delay_level_label: string
  actual_percent?: number | null
  planned_percent?: number | null
  progress_deviation?: number | null
  responsible_person?: string | null
  responsible_unit?: string | null
  planned_finish_date?: string | null
  status: RectificationStatus
  status_label: string
  review_result?: string | null
  remark?: string | null
  is_overdue: boolean
  created_at: string
  updated_at: string
  closed_at?: string | null
}

export type RectificationListResponse = {
  items: RectificationItem[]
  total: number
  page: number
  page_size: number
}

export type RectificationSummary = {
  total: number
  open: number
  in_progress: number
  completed: number
  closed: number
  ignored: number
  overdue: number
  serious: number
  new_this_week: number
  closed_this_week: number
}

export type RectificationActionLog = {
  id: number
  rectification_item_id: number
  project_id: number
  action: string
  action_label: string
  operator: string
  from_status?: string | null
  from_status_label?: string | null
  to_status?: string | null
  to_status_label?: string | null
  content?: string | null
  created_at: string
}

export type RectificationFilters = {
  page?: number
  page_size?: number
  sort_by?: string
  sort_order?: 'asc' | 'desc'
  status?: string
  delay_level?: string
  discipline?: string
  building?: string
  floor?: string
  responsible_person?: string
  responsible_unit?: string
  overdue?: boolean | null
  source_type?: string
  keyword?: string
  scope?: string | null
  data_date?: string | null
  import_group_id?: string | null
  batch_ids?: string | null
  batch_id?: number | null
  baseline_plan_id?: number | null
  calculation_method?: string | null
}

export type RectificationFilterOptions = {
  disciplines: string[]
  buildings: string[]
  floors: string[]
  responsible_persons: string[]
  responsible_units: string[]
  delay_levels: string[]
  statuses: string[]
  source_types: string[]
  floors_by_building: Record<string, string[]>
}
